src/CommunityToolkit.Aspire.Hosting.PowerShell/PowerShellScriptResource.cs

```diff
@@ -0,0 +1,269 @@
+        /// </summary>
+        /// <param name="scriptLogger"></param>
+        /// <param name="notificationService"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public async Task StartAsync(ILogger scriptLogger,
+            ResourceNotificationService notificationService,
+            CancellationToken cancellationToken = default)
+        {
+            Debug.Assert(scriptLogger != null);
+            _scriptLogger = scriptLogger;
+
+            Debug.Assert(_parent.Pool != null);
+            _ps.RunspacePool = _parent.Pool;
+
+            ConfigurePSDataStreams(scriptLogger, notificationService);
+
+            _ps.InvocationStateChanged += async (_, args) =>
+            {
+                var knownState = args.InvocationStateInfo.State switch
+                {
+                    PSInvocationState.NotStarted => KnownResourceStates.NotStarted,
+                    PSInvocationState.Running => KnownResourceStates.Running,
+                    PSInvocationState.Completed => KnownResourceStates.Finished,
+                    PSInvocationState.Stopped => KnownResourceStates.Exited,
+                    PSInvocationState.Failed => KnownResourceStates.FailedToStart,
+                    PSInvocationState.Stopping => KnownResourceStates.Stopping,
+                    _ => throw new ArgumentOutOfRangeException( // probably should be assertion
+                        nameof(args.InvocationStateInfo.State),
+                        args.InvocationStateInfo.State,
+                        "Unknown PowerShell invocation state")
+                };
+
+                await notificationService.PublishUpdateAsync(this,
+                    state =>
+                    {
+                        state = state with
+                        {
+                            State = knownState,
+                            Properties = [
+                                .. state.Properties,
+                                new( "PSInvocationState", args.InvocationStateInfo.State.ToString() ),
+                                new( "Reason", args.InvocationStateInfo.Reason?.Message ?? string.Empty ),
+                            ]
+                        };
+
+                        if (knownState == KnownResourceStates.Running)
+                        {
+                            state = state with
+                            {
+                                StartTimeStamp = DateTime.Now,
+                            };
+                        }
+
+                        if (KnownResourceStates.TerminalStates.Contains(knownState))
+                        {
+                            state = state with
+                            {
+                                StopTimeStamp = DateTime.Now,
+                            };
+                        }
+
+                        return state;
```

Comment:
We can simplify this a little with a switch expression and some pattern matching.

Reply:
Ah, so KnownResourceStates isn't an enum. Switch expressions want constants on the LHS. I simplified it nonetheless.